### PR TITLE
Use reduction producer for vectorization analysis

### PR DESCRIPTION
**Pull request by zasdfgbnm**
So that there is no special handling for reduction in `innerMostRootDim` and friends, improving code readability.

Checked cuda kernels with benchmarks, no change.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

File: csrc/scheduler/normalization.cpp [hunk 1]
```diff
@@ -1304,8 +1304,8 @@ std::shared_ptr<ReductionParams> getPersistentHeuristics(
       !persistent_buffer_info.persistent_buffers.empty(),
       "Persistent scheduler requires persistent buffers.");
 
-  auto properties =
-      scheduler_utils::getProperties(fusion, runtime_info, first_red_tv);
```
Comment (zasdfgbnm) [inline csrc/scheduler/normalization.cpp:1308]: Just a renaming, no behavioral change

File: csrc/scheduler/utils.cpp [hunk 2]
```diff
@@ -2005,8 +1991,7 @@ void propagateViewTransforms(Fusion* fusion, const ComputeAtMap& ca_map) {
 }
 
 bool isFastestDimReduction(TensorView* tv) {
-  for (auto it = tv->getMaybeRFactorDomain().rbegin();
-       it != tv->getMaybeRFactorDomain().rend();
+  for (auto it = tv->getRootDomain().rbegin(); it != tv->getRootDomain().rend();
```
Comment (zasdfgbnm) [inline csrc/scheduler/utils.cpp:1994]: I think using root domain is more aligned with the deleted comment:
```
  // This is backwards from how we normally think about grabbing root dimensions
  // to process. If we're in a reduction scheduler and we're using the rfactored
  // reduction tensor view, we don't care about the rfactor domain, we care
  // about the root domain because we're looking to vectorize the reads (input
  // tensor views). Otherwise we do want the rfactor domain. So this is the
  // reverse of our typical check, we actually want to selectively ignore the
  // rfactor domain.
```

File: csrc/scheduler/utils.cpp [hunk 3]
```diff
@@ -575,7 +575,7 @@ TvProperties getProperties(
   // Start from the inner most dimension, and work outwards. If this is a 3D
   // pattern, i.e. theres a pattern like [r0, r1, i2, r3] or [i0, r1, r2, i3,
   // i4] then compute the inner most dimension to compute separately.
-  const auto& root_dom = tv->getMaybeRFactorDomain();
+  const auto& root_dom = tv->getRootDomain();
```
Comment (zasdfgbnm) [inline csrc/scheduler/utils.cpp:578]: I think using root domain is more aligned with the deleted comment:
```C++
  // This is backwards from how we normally think about grabbing root dimensions
  // to process. If we're in a reduction scheduler and we're using the rfactored
  // reduction tensor view, we don't care about the rfactor domain, we care
  // about the root domain because we're looking to vectorize the reads (input
  // tensor views). Otherwise we do want the rfactor domain. So this is the
  // reverse of our typical check, we actually want to selectively ignore the
  // rfactor domain.
```

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
debug:
```
$PYTORCH_NVFUSER_DUMP="scheduler_params" ./bin/nvfuser_bench --benchmark_filter=NvFuserScheduler_LayerNorm_BWD_fp16/8192 --benchmark_min_time=0.000001
```

**Comment by zasdfgbnm** [comment]
!build

File: csrc/scheduler/utils.cpp [hunk 4]
```diff
@@ -1266,10 +1252,14 @@ void FindAllMappedDims::propagateSibling(TensorView* from, TensorView* to) {
...
-    mapped_id_set.emplace(entry.second);
+    if (entry.second != nullptr) {
+      mapped_id_set.emplace(entry.second);
+    }
   }
   for (auto entry : mapped_rfactor_ids_) {
-    mapped_id_set.emplace(entry.second);
+    if (entry.second != nullptr) {
+      mapped_id_set.emplace(entry.second);
+    }
```
Comment (zasdfgbnm) [inline csrc/scheduler/utils.cpp:1262]: There should be no behavior change here, I am just making sure that `mapped_id_set` does not contain `nullptr` so we can just use `ir_utils::toString` to print it.

File: csrc/scheduler/utils.cpp [hunk 5]
```diff
@@ -1181,12 +1172,7 @@ IterDomain* innerMostRootDim(TensorView* tv) {
   IterDomain* inner_most_id = nullptr;
 
   for (auto it = root_domain.rbegin(); it != root_domain.rend(); it++) {
-    // If we're looking at a reduction domain on an input because of
-    // segmentation we don't want to consider those reduction domains as a
-    // vectorization opportunity. If we're looking at a reduction reference
-    // tensor we want to consider the reduction iteration domains as domains we
-    // can vectorize on.
-    if (((*it)->isReduction() && tv->isFusionInput()) || (*it)->isBroadcast()) {
```
Comment (naoyam) [inline csrc/scheduler/utils.cpp:1189]: Why is this changed?

File: csrc/scheduler/utils.cpp (hunk 4 above)
Comment (naoyam) [reply csrc/scheduler/utils.cpp:1262]: Why can it be `nullptr`? Can we just ignore it?

File: csrc/scheduler/utils.cpp [hunk 6]
```diff
@@ -1279,8 +1269,7 @@ bool hasInnerDim(
     std::unordered_set<IterDomain*> inner_dims,
     bool should_vectorize) {
   const auto& inner_most_dim = innerMostRootDim(tv);
-  // TODO: Why "|| inner_most_dim->isReduction()"
```
Comment (naoyam) [inline csrc/scheduler/utils.cpp:1282]: `innerMostRootDim` should never return reduction domains? Maybe add an assertion just in case?

File: csrc/scheduler/utils.cpp (hunk 2 above)
Comment (naoyam) [reply csrc/scheduler/utils.cpp:1997]: Shouldn't this be the allocation domain because that defines which domains are fastest

File: csrc/scheduler/utils.cpp (hunk 3 above)
Comment (naoyam) [reply csrc/scheduler/utils.cpp:578]: IIUC, at this point, there should be no reduction rfactor, right? Should we assert that?

File: csrc/scheduler/utils.cpp (hunk 2 above)
Comment (zasdfgbnm) [reply csrc/scheduler/utils.cpp:1997]: Yes, I think so, I will change this to allocation domain, although I haven't started supporting allocation domain for schedulers yet so simply changing this to allocation domain does not make things better in practice.

File: csrc/scheduler/utils.cpp (hunk 4 above)
Comment (zasdfgbnm) [reply csrc/scheduler/utils.cpp:1262]: `nullptr` mean in the given tensor, nothing is mapped. With the current code organization, it is just slightly more convenient to put a `nullptr` to `mapped_rfactor_ids_` rather than not putting anything.

File: csrc/scheduler/utils.cpp (hunk 5 above)
Comment (zasdfgbnm) [reply csrc/scheduler/utils.cpp:1189]: This is exactly the purpose of this PR. In the past, this function was called both on the reduction tv and fusion inputs. So we need to special handle reduction tv. Now, this function will not be called on the reduction tv (instead, it will be the producer of the reduction tv), so there is no need to special handle anything.

File: csrc/scheduler/utils.cpp (hunk 3 above)
Comment (zasdfgbnm) [reply csrc/scheduler/utils.cpp:578]: Yes, there should be no reduction rfactor, but there could be transpose rfactor fused in the reduction, for the case of matmul. So I would like to leave the door open for having an rFactor.

File: csrc/scheduler/utils.cpp (hunk 6 above)
Comment (zasdfgbnm) [reply csrc/scheduler/utils.cpp:1282]: No, `innerMostRootDim` skips both broadcast and reduction now. `innerMostRootDim` is actually very simple:
```C++
IterDomain* innerMostRootDim(TensorView* tv) {
  const auto& root_domain = tv->getMaybeRFactorDomain();

  if (tv->nDims() == 0) {
    return nullptr;
  }

  IterDomain* inner_most_id = nullptr;

  for (auto it = root_domain.rbegin(); it != root_domain.rend(); it++) {
    if ((*it)->isReduction() || (*it)->isBroadcast()) {
      continue;
    }
    inner_most_id = *it;
    break;
  }

  return inner_most_id;
}
```
I prefer to not assert this, because:
1. This function is simple enough
2. We are not asserting that this function does not return a broadcast, so we should be consistent for reduction.

File: csrc/scheduler/utils.cpp (hunk 2 above)
Comment (zasdfgbnm) [reply csrc/scheduler/utils.cpp]: Changed

File: csrc/scheduler/utils.cpp (hunk 3 above)
Comment (naoyam) [reply csrc/scheduler/utils.cpp:578]: But transpose tensors should not have reduction domains, right?

Comment (zasdfgbnm) [reply csrc/scheduler/utils.cpp:578]: They can, for such a case, it is a fused transpose-reduction. An example for transpose-reduction is NN matmul, where the reduction tensor has root domain `[N, rK, M]` but rFactor domain `[M, N, rK]`

File: csrc/scheduler/utils.cpp (hunk 5 above)
Comment (zasdfgbnm) [reply csrc/scheduler/utils.cpp:1189]: Yes, exactly, and this is why we have all the special handling in these util functions. And these special handlings are hard to maintain and hard to understand, so I want to remove them.

File: csrc/scheduler/utils.cpp (hunk 3 above)
Comment (naoyam) [reply csrc/scheduler/utils.cpp:578]: OK, I forgot that's now possible. Makes sense.

**Review by naoyam — APPROVED** [review]
LGTM
